src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/DocumentSymbols/RemoteDocumentSymbolService.cs

```diff
@@ -43,9 +44,47 @@ protected override IRemoteDocumentSymbolService CreateService(in ServiceArgs arg
             supportsVSExtensions: _clientCapabilitiesService.ClientCapabilities.SupportsVisualStudioExtensions,
             cancellationToken).ConfigureAwait(false);
 
+        // Roslyn uses an internal "RoslynDocumentSymbol" type, which throws when serialized after we've mapped it, so we have to
+        // convert things back to DocumentSymbol. We only need to do the first level though, as our remapping will take care of
+        // the children.
+        if (csharpSymbols.TryGetFirst(out var roslynDocumentSymbols))
+        {
+            csharpSymbols = ConvertDocumentSymbols(roslynDocumentSymbols);
+        }
+
         var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
         var csharpDocument = codeDocument.GetCSharpDocument();
 
         return _documentSymbolService.GetDocumentSymbols(context.Uri, csharpDocument, csharpSymbols);
     }
+
+    [return: NotNullIfNotNull(nameof(roslynDocumentSymbols))]
+    private static DocumentSymbol[]? ConvertDocumentSymbols(DocumentSymbol[]? roslynDocumentSymbols)
```

Comment:
This might be controversial, but should this just replace in the original array rather than creating a new one?

Reply:
I'll try it, but I think the underlying type might be the Roslyn internal type, and I think the serializer is too smart.

Reply:
eh if it's a pain then np just leave it. It's such a small perf difference.

Reply:
Nope, didn't work. But the good news is, I've never seen this exception before!

```
System.ArrayTypeMismatchException : Attempted to access an element as a type incompatible with the array.
```